projects/example-app/src/app/auth/actions/auth-api.actions.ts

```diff
@@ -7,20 +7,12 @@ export enum AuthApiActionTypes {
   LoginRedirect = '[Auth/API] Login Redirect',
 }
 
-export class LoginSuccess implements Action {
-  readonly type = AuthApiActionTypes.LoginSuccess;
+export const AuthApiActions = {
```

Comment:
I think having functions directly in the file would be better. You can then use named module import as a grouping mechanism.

Reply:
We have to group them this way to take advantage of the ActionsUnion?

Reply:
So a few issues I have with this:
- both function and classes Action creator are allowed and welcomed. This changes to function-specific.
- according to Google's TS style guide:

> When providing a structural-based implementation, explicitly include the type at the declaration of the symbol (this allows more precise type checking and error reporting).

```typescript
// use
const foo: Foo = {
  a: 123,
  b: 'abc',
}
// instead of
const foo = {
  a: 123,
  b: 'abc',
}
```

On the other hand, I don't think anyone at Google is using function-based action creator and this proposal could be fine.

I'm planning to propose something like this for the class-based action creators: https://www.typescriptlang.org/play/index.html#src=%2F%2F%20action_helper.ts%0D%0Ainterface%20Type%3CC%2C%20T%20extends%20string%3E%20%7B%0D%0A%20%20%20new%20(...args%3A%20any%5B%5D)%3A%20C%3B%0D%0A%20%20%20readonly%20type%3A%20T%3B%0D%0A%7D%0D%0Aabstract%20class%20StaticAction%3CT%20extends%20string%3E%20%7B%0D%0A%20%20static%20readonly%20type%3A%20string%3B%0D%0A%20%20readonly%20type!%3A%20T%3B%0D%0A%7D%0D%0A%0D%0Afunction%20Action%3CT%20extends%20string%3E(type%3A%20T)%3A%20Type%3CStaticAction%3CT%3E%2C%20T%3E%20%7B%0D%0A%20%20return%20class%20extends%20StaticAction%3CT%3E%20%7B%0D%0A%20%20%20%20readonly%20type%20%3D%20type%3B%0D%0A%20%20%20%20static%20type%20%3D%20type%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20MyAction%20extends%20Action('foo')%20%7B%7D%0D%0A%0D%0Aconsole.log(new%20MyAction().type)%3B%20%2F%2F%20logs%3A%20'foo'%0D%0Aconsole.log(MyAction.type)%3B%20%2F%2F%20logs%3A%20'foo

Reply:
I should add that this was @kolodny idea, that I iterated on.

Reply:
Will you open a new issue with this proposal? I'd lean towards creating a new API such as `withAction('type')` for this as opposed to changing `Action`.